tests: Bluetooth: Mesh: improve test system clock precision

alxelax opened the pull request:
Increase system clock precision in the test system. It helps for some events happen in more precise time and prevents time overlapping between them.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf:
```diff
@@ -1,6 +1,6 @@
 CONFIG_LOG_MODE_IMMEDIATE=y
 CONFIG_ASSERT=y
-CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=10000
```
Should it be 32768?

alxelax replied:
32768 is too high. Some of beacon tests (multiple_netkeys.sh and beacon_interval.sh) start falling due to very high precision that slows down test. For instance multiple_netkeys.sh cannot send the last four beacons to pass tx scenario. Increasing test waiting time doesn't help. Test fails with internal reason 1(whatever it means) from main. Seems recalculation large waiting time with high precision overflows something within babblesim. I guess this is bug or limitation inside bsim environment. But yes for simple and short tests 32768 is better but we have large and complex tests as well :(

Andrewpini replied:
Do you know if the increased precision affects the general runtime of the tests? (Ref: Some bsim tests takes too long time to execute)

alxelax replied:
Yep, it makes the runtime of the tests a bit longer. However, it is not visible `on the eye` for the current tests. Nevertheless, the longest tests in the current set of tests (beacon tests) can still fit within `WAIT_TIME` definition. Unfortunately, they cannot if increase precision till 32768.
If to not do this then tests like reprovisioning with further health model configuration become not passable. Since leftover network transmission coincides with retransmission etc etc. It doesn't matter in which direction to move that and even 1 second delay is not enough to prevent this. Babblesim environment with low precision doesn't follow `xmit` settings and some frames can occur quite far from an expected interval. Additionally to the legacy advertiser bug, it gives unstable behavior for tests with significant traffic. 
I saw this already twice. The first time was when I made extended advertiser as default the second time was here.

PavelVPV replied:
Hmm, multiple_netkeys.sh does not fail on my desk if I set timeout to 640 seconds:
```
d_00: @00:05:12.356080  [00:05:12.356,079] <wrn> test_beacon: Didn't receive SNB in time (err: -11)
d_01: @00:05:20.022797 INFO: test_rx_multiple_netkeys PASSED
d_00: @00:05:21.235871  [00:05:21.235,870] <inf> test_beacon: Advertising started
d_00: @00:05:21.235871  
d_00: @00:05:30.080817  [00:05:30.080,810] <inf> test_beacon: Received beacon for Net Key Idx 1
d_00: @00:05:30.108405  [00:05:30.108,398] <inf> test_beacon: Received beacon for Net Key Idx 1
d_00: @00:05:30.137030  [00:05:30.137,023] <inf> test_beacon: Received beacon for Net Key Idx 2
d_00: @00:05:30.164954  [00:05:30.164,947] <inf> test_beacon: Received beacon for Net Key Idx 2
d_00: @00:05:30.206854  [00:05:30.206,848] <inf> test_beacon: Received beacon for Net Key Idx 3
d_00: @00:05:30.237158  [00:05:30.237,152] <inf> test_beacon: Received beacon for Net Key Idx 4
d_00: @00:05:40.082160  [00:05:40.082,153] <inf> test_beacon: Received beacon for Net Key Idx 1
d_00: @00:05:40.110297  [00:05:40.110,290] <inf> test_beacon: Received beacon for Net Key Idx 1
d_00: @00:05:40.134955  [00:05:40.134,948] <inf> test_beacon: Received beacon for Net Key Idx 2
d_00: @00:05:40.163611  [00:05:40.163,604] <inf> test_beacon: Received beacon for Net Key Idx 2
d_00: @00:05:40.180609  [00:05:40.180,603] <inf> test_beacon: Received beacon for Net Key Idx 3
d_00: @00:05:40.210578  [00:05:40.210,571] <inf> test_beacon: Received beacon for Net Key Idx 3
d_00: @00:05:40.230597  [00:05:40.230,590] <inf> test_beacon: Received beacon for Net Key Idx 4
d_00: @00:05:41.230622 INFO: test_tx_multiple_netkeys PASSED
```

beacon_interval.sh has actual failure:
```
d_01: @00:07:53.029786  main: The TESTCASE FAILED with return code 2
d_01: @00:07:53.029786 ERROR: (WEST_TOPDIR/zephyr/tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_beacon.c:913): delta >= (600 * MSEC_PER_SEC) is false.
```

And `delta` is 390004. I guess this should be investigated?

alxelax replied:
The same options(CONFIG_SYS_CLOCK_TICKS_PER_SEC=32768 and timeout 640 seconds) but on my PC:
```
d_00: @00:05:12.356080  [00:05:12.356,079] <wrn> test_beacon: Didn't receive SNB in time (err: -11)
d_01: @00:05:20.022797 INFO: test_rx_multiple_netkeys PASSED
d_00: @00:05:21.235871  [00:05:21.235,870] <inf> test_beacon: Advertising started
d_00: @00:05:21.235871  
d_00: @00:05:30.080817  [00:05:30.080,810] <inf> test_beacon: Received beacon for Net Key Idx 1
d_00: @00:05:30.108405  [00:05:30.108,398] <inf> test_beacon: Received beacon for Net Key Idx 1
d_00: @00:05:30.137030  [00:05:30.137,023] <inf> test_beacon: Received beacon for Net Key Idx 2
d_00: @00:05:30.164954  [00:05:30.164,947] <inf> test_beacon: Received beacon for Net Key Idx 2
d_00: @00:05:30.206854  [00:05:30.206,848] <inf> test_beacon: Received beacon for Net Key Idx 3
d_00: @00:05:30.237158  [00:05:30.237,152] <inf> test_beacon: Received beacon for Net Key Idx 4
d_00: @00:05:39.953563  main: The TESTCASE FAILED with return code 1
```
I'm pretty sure there is some bug or limitation in BabbleSim. It is not worth to use such precision with high waiting time.

PavelVPV replied:
Can you run the test with gdb?

alxelax replied:
I figured out. The test length should be the same for each emulated device. Otherwise, this issue happens. Nevertheless, why 32768 is better than 10000?

PavelVPV replied:
Probably not `100000`, but `10000` (what is set in this PR). 32768 is the same value as set on physical devices. And I also expected better precision especially in tests where the time matters, e.g. where we test model publication or beacons.

alxelax replied:
Hm. I didn't mention `100000` anywhere. I fixed beacon tests for 32768. Additionally, I found the bug in bsim settings backend. It can lose lines if the name of the new setting is the occurrence of any existing ones.

PavelVPV commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/settings_test_backend.c:
```diff
@@ -39,9 +39,12 @@ static char setting_file_tmp[sizeof(setting_file) + 1];
 	char line[READ_LEN_MAX + 1];
+	char name_tmp[strlen(name) + 2];
+
+	snprintk(name_tmp, sizeof(name_tmp), "%s=", name);
```
Could you please move the fix to a separate commit?

alxelax committed:
tests: Bluetooth: Mesh: fix bsim settings backend

PR fixes bsim settings backend issue.
Backend looses line if setting name of new line is
the occurence of the any existing lines.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>

alxelax committed:
tests: Bluetooth: Mesh: improve test system clock precision

Increase system clock precision in the test system.
It helps for some events happen in more precise time and
prevents time overlapping between them.

Signed-off-by: Aleksandr Khromykh <aleksandr.khromykh@nordicsemi.no>
